test/unit/test-builder.js

```diff
@@ -0,0 +1,413 @@
+  });
+
+  afterEach(() => {
+    clock.uninstall();
+  });
+
+  function createAmpElement(options = {}) {
+    const element = createElementWithAttributes(doc, 'amp-el', {});
+    element.deferredBuild = () => options.deferredBuild || false;
+    element.prerenderAllowed = () => options.prerenderAllowed || false;
+    element.getBuildPriority = () =>
+      options.buildPriority || LayoutPriority.CONTENT;
+    element.buildInternal = env.sandbox.stub();
+    return element;
+  }
+
+  describe('schedule', () => {
+    it('should schedule a deferredBuild element', () => {
+      const element = createAmpElement({deferredBuild: true});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.true;
+
+      builder.unschedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should schedule a non-deferredBuild element', () => {
+      const element = createAmpElement({deferredBuild: false});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should unschedule when built', async () => {
+      const element = createAmpElement({deferredBuild: true});
+      builder.schedule(element);
+      expect(intersectionObserverStub.isObserved(element)).to.be.true;
+
+      await setAmpdocReady();
+      intersectionObserverStub.notifySync({
+        target: element,
+        isIntersecting: true,
+      });
+      expect(intersectionObserverStub.isObserved(element)).to.be.false;
+    });
+
+    it('should signal READY_SCAN_SIGNAL after first element scheduled', async () => {
+      ampdoc.signals().reset(READY_SCAN_SIGNAL);
+      const element = createAmpElement({deferredBuild: false});
+      builder.schedule(element);
+      expect(ampdoc.signals().get(READY_SCAN_SIGNAL)).to.be.null;
+
+      clock.tick(50);
+      expect(ampdoc.signals().get(READY_SCAN_SIGNAL)).to.exist;
+    });
+
+    it('should signal READY_SCAN_SIGNAL after document ready', async () => {
+      ampdoc.signals().reset(READY_SCAN_SIGNAL);
+      await setAmpdocReady();
+      clock.tick(50);
+      expect(ampdoc.signals().get(READY_SCAN_SIGNAL)).to.exist;
+    });
+  });
+
+  describe('wait for parsing', () => {
+    it('should build when document ready', async () => {
+      await setAmpdocReady();
+      const element = createAmpElement({deferredBuild: false});
+      builder.schedule(element);
+      clock.tick(1);
+      expect(element.buildInternal).to.be.calledOnce;
+    });
+
+    it('should build when document becomes ready', async () => {
+      const element = createAmpElement({deferredBuild: false});
+      builder.schedule(element);
+      clock.tick(1);
+      expect(element.buildInternal).to.be.not.called;
+
+      await setAmpdocReady();
+      clock.tick(1);
+      expect(element.buildInternal).to.be.calledOnce;
+    });
+
+    it('should build asap when document ready', async () => {
```

Comment:
Nit: Do you have a test for `'should wait for parse even for asap scheduled element'`

Reply:
Yes. All "should be asap" check that `buildInternal` has not been called at first and only called after parse complete has been determined.